File: LeetCode/Problems/Python/987. Vertical Order Traversal of a Binary Tree.py
#Partial sorting with hashtable
#Time: O(nlog(n/k)), we have k cols, so we will do k sorting on each of (n/k)log(n/k) --> nlog(n/k), slightly faster than global sorting
#Space: O(n)
class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return None
        
        queue = [(root, (0,0))]
        column_table = defaultdict(list)
        min_col = float('inf')
        max_col = float('-inf')
        while queue:
            for _ in range(len(queue)):
                cur, (col, row) = queue.pop(0)
                column_table[col].append((row,cur.val))
                min_col = min(min_col, col)
                max_col = max(max_col, col)
                if cur.left:
                    queue.append((cur.left, (col-1, row+1)))
                if cur.right:
                    queue.append((cur.right, (col+1, row+1)))
        
        return [[val for row, val in sorted(column_table[col])] for col in range(min_col, max_col+1)]

#BFS + Global sorting
#Time: O(nlogn)
#Space: O(n)
class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return None
        
        queue = [(root, 0, 0)]
        nodes = []
        while queue:
            for _ in range(len(queue)):
                cur, col, row = queue.pop(0)
                nodes.append((col, row, cur.val))
                if cur.left:
                    queue.append((cur.left, col-1, row+1))
                if cur.right:
                    queue.append((cur.right, col+1, row+1))
                    
        nodes.sort()
        ans = defaultdict(list)
        for node in nodes:
            ans[node[0]].append(node[2])
            
        return ans.values()

File: LeetCode/Problems/C++/987. Vertical Order Traversal of a Binary Tree.cpp
//Time: O(nlog(n/k))
//Space: O(n)
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        if(root == NULL) return {};
        vector<vector<int>> vertical_order;
        
        queue<pair<TreeNode*,pair<int,int>>> myqueue;
        
        myqueue.push({root, {0,0}});
        int min_col = INT_MAX;
        int max_col = INT_MIN;
        int level_size;
        
        unordered_map<int, vector<pair<int,int>>> column_table;
        while(!myqueue.empty()){
            level_size = myqueue.size();
            for(int i=0;i<level_size;i++){
                TreeNode* curr = myqueue.front().first;
                int col = myqueue.front().second.first;
                int row = myqueue.front().second.second;
                myqueue.pop();
                if(curr==NULL) continue;
                column_table[col].push_back({row, curr->val});
                max_col = max(max_col, col);
                min_col = min(min_col, col);
                myqueue.push({curr->left, {col-1, row+1}});
                myqueue.push({curr->right, {col+1, row+1}});
            }
        }
        
        for(int i=min_col;i<=max_col;i++){
            sort(column_table[i].begin(), column_table[i].end());
            vector<int> temp;
            for(int j=0;j<column_table[i].size();j++)
                temp.push_back(column_table[i][j].second);
            vertical_order.push_back(temp);
        }
        
        return vertical_order;
        
    }
};